implementations/utils/gpu.py

```diff
@@ -0,0 +1,394 @@
+      delete_protection: boolean value indicating if the new virtual machine should be
+          protected against deletion or not.
+  Returns:
+      Instance object.
+  """
+  instance_client = compute_v1.InstancesClient()
+
+  # Use the network interface provided in the network_link argument.
+  network_interface = compute_v1.NetworkInterface()
+  network_interface.network = network_link
+  if subnetwork_link:
+    network_interface.subnetwork = subnetwork_link
+
+  if internal_ip:
+    network_interface.network_i_p = internal_ip
+
+  if external_access:
+    access = compute_v1.AccessConfig()
+    access.type_ = compute_v1.AccessConfig.Type.ONE_TO_ONE_NAT.name
+    access.name = "External NAT"
+    access.network_tier = access.NetworkTier.PREMIUM.name
+    if external_ipv4:
+      access.nat_i_p = external_ipv4
+    network_interface.access_configs = [access]
+
+  # Collect information into the Instance object.
+  instance = compute_v1.Instance()
+  instance.network_interfaces = [network_interface]
+  instance.name = instance_name
+  instance.disks = disks
+  if re.match(r"^zones/[a-z\d\-]+/machineTypes/[a-z\d\-]+$", machine_type):
+    instance.machine_type = machine_type
+  else:
+    instance.machine_type = f"zones/{zone}/machineTypes/{machine_type}"
+
+  instance.scheduling = compute_v1.Scheduling()
+  if accelerators:
+    instance.guest_accelerators = accelerators
+    instance.scheduling.on_host_maintenance = (
+        compute_v1.Scheduling.OnHostMaintenance.TERMINATE.name
+    )
+
+  if metadata:
+    instance.metadata = metadata
+
+  if service_account:
+    instance.service_accounts = [service_account]
+
+  if preemptible:
+    # Set the preemptible setting
+    logging.warn("Preemptible VMs are being replaced by Spot VMs.", DeprecationWarning)
+    instance.scheduling = compute_v1.Scheduling()
+    instance.scheduling.preemptible = True
+
+  if spot:
+    # Set the Spot VM setting
+    instance.scheduling.provisioning_model = (
+        compute_v1.Scheduling.ProvisioningModel.SPOT.name
+    )
+    instance.scheduling.instance_termination_action = instance_termination_action
+
+  if custom_hostname is not None:
+    # Set the custom hostname for the instance
+    instance.hostname = custom_hostname
+
+  if delete_protection:
+    # Set the delete protection bit
+    instance.deletion_protection = True
+
+  # Prepare the request to insert an instance.
+  request = compute_v1.InsertInstanceRequest()
+  request.zone = zone
+  request.project = project_id
+  request.instance_resource = instance
+
+  # Wait for the create operation to complete.
+  logging.info(f"Creating the {instance_name} instance in {zone}...")
+
+  operation = instance_client.insert(request=request)
+
+  wait_for_extended_operation(operation, "instance creation")
+
+  logging.info(f"Instance {instance_name} created.")
+  return instance_client.get(project=project_id, zone=zone, instance=instance_name)
+
+
+def wait_for_extended_operation(
```

Comment:
nit: this can be a `task.sensor` like `wait_for_queued_resource`

Reply:
After adding `@task.sensor(poke_interval=60, timeout=3600, mode="reschedule")` for the function, the GPU VM is created correctly but the task is still marked to retry.  And the second time we retry build, it will error out complaining the tpu name already exists.

Reply:
Have you updated the return value? The return value of the sensor needs to be a bool. https://github.com/GoogleCloudPlatform/ml-auto-solutions/blob/master/implementations/utils/tpu.py#L118